### ISSUE TITLE
Dart2JS crash: Runtime type information not available for type_variable_local in (Instance of 'ThisLocalVariable') for generative_constructor_body.

### ISSUE THREAD

**Issue by lexaknyazev** [dart-lang/sdk#29722]
**Dart SDK 1.24.0-dev.6.0**

```dart
class Foo {
  Foo() {
    bar<T>(T function()) {}
  }
}

void main() {
  new Foo();
  print(null.runtimeType);
}
```

Dart2JS output:
```
Transform Dart2JS on bug_demo|web/main.dart threw error: Assertion failure: Cannot find value Instance of 'TypeVariableLocal' in (Instance of 'ThisLocal').
package:compiler_unsupported/src/ssa/locals_handler.dart 330           LocalsHandler.readLocal
package:compiler_unsupported/src/ssa/builder.dart 1880                 SsaBuilder.visitFunctionExpression.<fn>
dart:core                                                              List.forEach
package:compiler_unsupported/src/ssa/builder.dart 1876                 SsaBuilder.visitFunctionExpression
package:compiler_unsupported/src/tree/nodes.dart 1132                  FunctionExpression.accept
package:compiler_unsupported/src/ssa/builder.dart 1587                 SsaBuilder.visit
package:compiler_unsupported/src/ssa/builder.dart 1891                 SsaBuilder.visitFunctionDeclaration
package:compiler_unsupported/src/tree/nodes.dart 1050                  FunctionDeclaration.accept
package:compiler_unsupported/src/ssa/builder.dart 1587                 SsaBuilder.visit
package:compiler_unsupported/src/ssa/builder.dart 1633                 SsaBuilder.visitBlock
package:compiler_unsupported/src/tree/nodes.dart 903                   Block.accept
package:compiler_unsupported/src/ssa/builder.dart 787                  SsaBuilder.buildMethod
package:compiler_unsupported/src/ssa/builder.dart 340                  SsaBuilder.build
package:compiler_unsupported/src/ssa/builder.dart 139                  SsaAstBuilderTask.build.<fn>.<fn>
package:compiler_unsupported/src/compiler.dart 1098                    CompilerDiagnosticReporter.withCurrentElement
package:compiler_unsupported/src/ssa/builder.dart 130                  SsaAstBuilderTask.build.<fn>
package:compiler_unsupported/src/common/tasks.dart 63                  CompilerTask.measure
package:compiler_unsupported/src/ssa/builder.dart 124                  SsaAstBuilderTask.build
package:compiler_unsupported/src/ssa/ssa.dart 37                       SsaFunctionCompiler.compile
package:compiler_unsupported/src/js_backend/backend.dart 932           JavaScriptBackend.codegen
package:compiler_unsupported/src/js_backend/element_strategy.dart 125  ElementCodegenWorkItem.run
package:compiler_unsupported/src/compiler.dart 737                     Compiler.emptyQueue.<fn>.<fn>.<fn>.<fn>.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 737                     Compiler.emptyQueue.<fn>.<fn>.<fn>.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 735                     Compiler.emptyQueue.<fn>.<fn>.<fn>
package:compiler_unsupported/src/compiler.dart 1098                    CompilerDiagnosticReporter.withCurrentElement
package:compiler_unsupported/src/compiler.dart 733                     Compiler.emptyQueue.<fn>.<fn>
package:compiler_unsupported/src/enqueue.dart 507                      EnqueuerStrategy.processWorkItem
package:compiler_unsupported/src/js_backend/enqueuer.dart 227          CodegenEnqueuer.forEach
package:compiler_unsupported/src/compiler.dart 729                     Compiler.emptyQueue.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 728                     Compiler.emptyQueue
package:compiler_unsupported/src/compiler.dart 752                     Compiler.processQueue.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 747                     Compiler.processQueue
package:compiler_unsupported/src/compiler.dart 621                     Compiler.compileLoadedLibraries.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 518                     Compiler.compileLoadedLibraries
package:compiler_unsupported/src/compiler.dart 473                     Compiler.runInternal
===== asynchronous gap ===========================
package:compiler_unsupported/src/compiler.dart 308                     Compiler.run.<fn>.<fn>
dart:async                                                             new Future.sync
package:compiler_unsupported/src/compiler.dart 308                     Compiler.run.<fn>
package:compiler_unsupported/src/common/tasks.dart 176                 CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 305                     Compiler.run
package:compiler_unsupported/src/apiimpl.dart 263                      CompilerImpl.run.<fn>.<fn>
Build failed.
```

Build succeeds if there's no call to `runtimeType`.

**Comment by sigmundch** [comment]
Thanks for the small repro!

+ @johnniwinther @rakudrama @eernstg  who are more familiar with this part of the system. Seems like we are crashing when using a generic type in a local closure.

@lexaknyazev - you might already know this, just wanted to note that you can probably workaround this if you move the closure out (e.g. make it a private method instead)

**Comment by lexaknyazev** [comment]
> Seems like we are crashing when using a generic type in a local closure.

Crash happens when all four conditions are true: 
- local function is defined within constructor;
- local function is generic;
- local function has parameter that uses generic type;
- `.runtimeType` is used somewhere in sources.

E.g., all following examples break compiler
```dart
class Foo {
  Foo() {
    bar<T>(T param) {}
  }
}
```
```dart
class Foo {
  Foo() {
    bar<T>(T function()) {}
  }
}
```
```dart
class Foo {
  Foo() {
    bar<T>(List<T> param) {}
  }
}
```
```dart
class Foo {
  Foo() {
    bar<T>(function(List<T> param)) {}
  }
}
```

**Comment by lexaknyazev** [comment]
@sigmundch 
Is there any chance of fixing this in 1.24 or 1.25?

**Comment by sigmundch** [comment]
Hey @lexaknyazev  - We unfortunately won't get to it for 1.24, but I'm also not sure we'll get to this by 1.25. Let us know if this is a blocking issue for you, though. We were prioritizing some other work based on the assumption that you can workaround this issue without too much trouble.

**Comment by lexaknyazev** [comment]
Since I'm not relying on `.runtimeType` getter in production code, it's not a blocking issue for me.
However, there could be other developers who may spend lots of time figuring out what's wrong with their code. E.g., in my case, `.runtimeType` and generic closure were in different libraries.

**Comment by lexaknyazev** [comment]
`Dart VM version: 2.0.0-dev.32.0`

```
Runtime type information not available for type_variable_local(bar.T) in (Instance of 'ThisLocalVariable') for generative_constructor_body(Foo#).
  Foo() {
  ^^^
The compiler is broken.

When compiling the above element, the compiler crashed. It is not
possible to tell if this is caused by a problem in your program or
not. Regardless, the compiler should not crash.

The Dart team would greatly appreciate if you would take a moment to
report this problem at http://dartbug.com/new.

Please include the following information:

* the name and version of your operating system,

* the Dart SDK build number (build number could not be determined), and

* the entire message you see here (including the full stack trace
  below as well as the source location above).

Build error:
Transform Dart2JS on bug_demo|web/main.dart threw error: Assertion failure: Runtime type information not available for type_variable_local(bar.T) in (Instance of 'ThisLocalVariable') for generative_constructor_body(Foo#).
package:compiler_unsupported/src/diagnostics/invariant.dart 55            failedAt
package:compiler_unsupported/src/ssa/locals_handler.dart 328              LocalsHandler.readLocal
package:compiler_unsupported/src/ssa/builder.dart 1922                    SsaAstGraphBuilder.visitFunctionExpression.<fn>
package:compiler_unsupported/src/elements/common.dart 388                 ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.forEachInstanceField.fieldFilter
package:compiler_unsupported/src/elements/common.dart 356                 ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.forEachMember.<fn>
package:front_end/src/fasta/util/link_implementation.dart 124             LinkEntry.forEach
package:compiler_unsupported/src/elements/modelx.dart 2985                ClassElementX.forEachLocalMember
package:compiler_unsupported/src/elements/common.dart 356                 ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.forEachMember
package:compiler_unsupported/src/elements/common.dart 392                 ElementX&AstElementMixin&AnalyzableElementX&ClassElementCommon.forEachInstanceField
package:compiler_unsupported/src/universe/codegen_world_builder.dart 605  ElementCodegenWorldBuilderImpl.forEachInstanceField
package:compiler_unsupported/src/ssa/builder.dart 1919                    SsaAstGraphBuilder.visitFunctionExpression
package:compiler_unsupported/src/tree/nodes.dart 1209                     FunctionExpression.accept
package:compiler_unsupported/src/ssa/builder.dart 1624                    SsaAstGraphBuilder.visit
package:compiler_unsupported/src/ssa/builder.dart 1933                    SsaAstGraphBuilder.visitFunctionDeclaration
package:compiler_unsupported/src/tree/nodes.dart 1127                     FunctionDeclaration.accept
package:compiler_unsupported/src/ssa/builder.dart 1624                    SsaAstGraphBuilder.visit
package:compiler_unsupported/src/ssa/builder.dart 1670                    SsaAstGraphBuilder.visitBlock
package:compiler_unsupported/src/tree/nodes.dart 980                      Block.accept
package:compiler_unsupported/src/ssa/builder.dart 769                     SsaAstGraphBuilder.buildMethod
package:compiler_unsupported/src/ssa/builder.dart 319                     SsaAstGraphBuilder.build
package:compiler_unsupported/src/ssa/builder.dart 114                     SsaAstBuilder.build.<fn>.<fn>
package:compiler_unsupported/src/compiler.dart 1156                       CompilerDiagnosticReporter.withCurrentElement
package:compiler_unsupported/src/ssa/builder.dart 105                     SsaAstBuilder.build.<fn>
package:compiler_unsupported/src/common/tasks.dart 63                     CompilerTask.measure
package:compiler_unsupported/src/ssa/builder.dart 99                      SsaAstBuilder.build
package:compiler_unsupported/src/ssa/ssa.dart 91                          SsaBuilderTask.build
package:compiler_unsupported/src/ssa/ssa.dart 43                          SsaFunctionCompiler.compile
package:compiler_unsupported/src/js_backend/backend.dart 848              JavaScriptBackend.codegen
package:compiler_unsupported/src/js_backend/element_strategy.dart 176     ElementCodegenWorkItem.run
package:compiler_unsupported/src/compiler.dart 784                        Compiler.emptyQueue.<fn>.<fn>.<fn>.<fn>.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 784                        Compiler.emptyQueue.<fn>.<fn>.<fn>.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 782                        Compiler.emptyQueue.<fn>.<fn>.<fn>
package:compiler_unsupported/src/compiler.dart 1156                       CompilerDiagnosticReporter.withCurrentElement
package:compiler_unsupported/src/compiler.dart 780                        Compiler.emptyQueue.<fn>.<fn>
package:compiler_unsupported/src/enqueue.dart 511                         EnqueuerStrategy.processWorkItem
package:compiler_unsupported/src/js_backend/enqueuer.dart 224             CodegenEnqueuer.forEach
package:compiler_unsupported/src/compiler.dart 776                        Compiler.emptyQueue.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 775                        Compiler.emptyQueue
package:compiler_unsupported/src/compiler.dart 797                        Compiler.processQueue.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 794                        Compiler.processQueue
package:compiler_unsupported/src/compiler.dart 640                        Compiler.compileLoadedLibraries.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 534                        Compiler.compileLoadedLibraries
package:compiler_unsupported/src/compiler.dart 488                        Compiler.runInternal
===== asynchronous gap ===========================
package:compiler_unsupported/src/compiler.dart 313                        Compiler.run.<fn>.<fn>
dart:async                                                                new Future.sync
package:compiler_unsupported/src/compiler.dart 313                        Compiler.run.<fn>
package:compiler_unsupported/src/common/tasks.dart 178                    CompilerTask.measureSubtask
package:compiler_unsupported/src/compiler.dart 310                        Compiler.run
package:compiler_unsupported/src/apiimpl.dart 265                         CompilerImpl.run.<fn>.<fn>
Build failed.
```

**Comment by lexaknyazev** [comment]
This bug seems to be fixed when using CFE as of `2.0.0-dev.39.0`

**Comment by sigmundch** [comment]
Thanks for following up. Yes, precisely. Support for generic methods is only available in the dev channel when using the new common front-end. The types in generic methods are only used when enabling `--strong`, but there are still a lot of issues that need to be fixed with that flag (it is likely to crash on simple programs today)
